### PR TITLE
Actually pass through the request for --color=always

### PR DESCRIPTION
**Pull request by saethlin**
https://github.com/rust-lang/miri/pull/2243 actually doesn't work :joy:

The suggestion to split on `,` was good but `arg` is actually the whole `--json=diagnostic-rendered-ansi,artifacts,future-incompat
`, and of course I didn't test that change locally and we have no test for this in CI.

Therefore, I would like some guidance on making a test for this because I'm going to rely on this working.

### CODE REVIEW

**Comment by oli-obk** [comment]
Basically create any test that emits a diagnostic. Then add `// compile-flags: your flags go here` to the file. Add only the json color flag, we already pass the flag for json diagnostics. Then bless and 🤞

**Comment by saethlin** [comment]
I did this:
```
$ cat tests/fail/color-always.rs 
// compile-flags: --color=always
// error-pattern: entering unreachable code
fn main() {
    unsafe { std::hint::unreachable_unchecked() }
}
```
But what I get out of `./miri test` looks like this, no evidence of ANSI escapes:
```
tests/fail/color-always.rs FAILED
command: "/home/ben/miri/target/x86_64-unknown-linux-gnu/release/miri" "--edition" "2018" "-Dwarnings" "-Dunused" "--sysroot" "/home/ben/.cache/miri/HOST" "-Zui-testing" "tests/fail/color-always.rs" "--error-format=json" "--color=always"

actual output differed from expected tests/fail/color-always.stderr
Diff < left / right > :
>error: Undefined Behavior: entering unreachable code
>  --> RUSTLIB/core/src/hint.rs:LL:CC
>   |
>LL |     unsafe { intrinsics::unreachable() }
>   |              ^^^^^^^^^^^^^^^^^^^^^^^^^ entering unreachable code
>   |
>   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
>   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
>           
>   = note: inside `std::hint::unreachable_unchecked` at RUSTLIB/core/src/hint.rs:LL:CC
>note: inside `main` at $DIR/color-always.rs:LL:CC
>  --> $DIR/color-always.rs:LL:CC
>   |
>LL |     unsafe { std::hint::unreachable_unchecked() }
>   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>
>note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
>
>error: aborting due to previous error
>
>
```

**Comment by saethlin** [comment]
Argh this patch also doesn't cover doctests. I'll look into that tomorrow.

**Comment by RalfJung** [comment]
The change is in cargo-miri, which is literally not called by the compiletest suite. So it is impossible to test this from the compiletest suite.

File: cargo-miri/bin.rs [hunk 1]
```diff
@@ -949,7 +949,9 @@ fn phase_runner(binary: &Path, binary_args: env::Args, phase: RunnerPhase) {
             // to emit the diagnostic structure that Cargo would consume from rustc to emit colored
             // diagnostics, and ask rustc to emit them.
             // See https://github.com/rust-lang/miri/issues/2037
-            if arg.split(',').any(|a| a == "diagnostic-rendered-ansi") {
+            // First skip over the leading `=`, then check for diagnostic-rendered-ansi in the
+            // comma-separated list
+            if suffix[1..].split(',').any(|a| a == "diagnostic-rendered-ansi") {
```
Comment (RalfJung) [inline cargo-miri/bin.rs:954]: Please do `let suffix = suffix.strip_prefix('=').unwrap()` instead of subslicing.

File: cargo-miri/bin.rs [hunk 2]
```diff
@@ -949,7 +949,10 @@ fn phase_runner(binary: &Path, binary_args: env::Args, phase: RunnerPhase) {
             // to emit the diagnostic structure that Cargo would consume from rustc to emit colored
             // diagnostics, and ask rustc to emit them.
             // See https://github.com/rust-lang/miri/issues/2037
-            if arg.split(',').any(|a| a == "diagnostic-rendered-ansi") {
+            // First skip over the leading `=`, then check for diagnostic-rendered-ansi in the
+            // comma-separated list
+            if suffix.strip_prefix('=').unwrap().split(',').any(|a| a == "diagnostic-rendered-ansi")
```
Comment (RalfJung) [inline cargo-miri/bin.rs:954]: This makes the `assert!(suffix.starts_with('='));` above redundant.

I think it's best to put the `let suffix = suffix.strip_prefix('=').unwrap()` where the `assert` is, also to keep this line a bit shorter.

**Comment by RalfJung** [comment]
FWIW I'd be fine with landing this as-is (assuming you tested it locally), and figuring out with a way to test this later.

**Comment by oli-obk** [comment]
Yea, I have plans for testing cargo miri, just so that we have a reasonable way to test with dependencies

**Comment by RalfJung** [comment]
So @saethlin could you resolve the last comment and mark this as ready? We can do the test part later.

**Comment by saethlin** [comment]
Just looked at this again and it seems like something is still missing. I'm running some tests, redirecting stdout and stderr to a file, and then `cat`ing the file. Those `ok`s should be green.
2022-06-25-113734_1278x572_scrot

**Comment by RalfJung** [comment]
That's not Miri/rustc output, that is  output from the application running inside Miri.

`--color=always` should only affect the warning/error diagnostics, I think.

**Comment by saethlin** [comment]
Hunh. To get fully colored output from `cargo test` you need to do `cargo test --color=always -- --color=always`.

But for cargo-miri, `cargo miri test --color=always -- --color=always` doesn't seem to pass the argument to the test programs. I tried throwing in extra `-- --color=always` but still can't get any green `ok`.

**Comment by RalfJung** [comment]
Miri blocks the `TERM` env var, which might lead libtest to never use colors.

But anyway that seems orthogonal to this PR?

**Comment by saethlin** [comment]
:shrug: perhaps it is. I'm just trying to get colored output from cargo-miri without lying to the process that it's connected to a TTY.

I'll poke around this later.

File: cargo-miri/bin.rs (hunk 2 above)
Comment (RalfJung) [reply cargo-miri/bin.rs:962]: This comment is still open.

**Comment by RalfJung** [comment]
>  Those oks should be green.

FWIW these are not green even when running this on a tty. So that is a different problem. And it's probably due to Miri removing the `TERM` env var.

**Pull request merged**